text/3579-pki.md

```diff
@@ -0,0 +1,262 @@
+[summary]: #summary
+
+This RFC proposes a plan and deployment for generating an initial set of Rust Project root and delegated cryptographic public/private key pairs, with associated X.509 certificate chains, to provide the proper hierarchy for deploying various signature validation techniques across the project.
+
+Having an internal Certificate Authority and delegate keys for the Rust Project allows us to have the infrastructure in place for solving multiple problems currently in the ecosystem such as: signing mirrors, crate package signing and validation, signing binaries for OS security checks, infrastructure security, and future chains of trust for authors and organizations.
+
+We propose the initial creation of a Rust Project Public Key hierarchy, with a root certificate authority and an initial set of single-purpose delegated certificates. For safety and oversight, this infrastructure mandates a quorum scheme (5-of-9) for root key operations, and mandates the use of Certificate Transparency for all subkey issuance *and* individual signature operations.
+
+Finally, we propose an initial, basic architecture for using this public key infrastructure to create delegate keys, as well as signing additional keys for usage in other signing schemas and implementations. Note, these additional implementations and further work is out of scope of this RFC. This RFC strictly covers creating a singular, root Certificate Authority and delegate keys for the Rust Project.
+
+## Out of Scope
+
+This RFC does not attempt to solve:
+- Identities or signing of Rust Project members
+- Identities or signing of crate authors
+- Answering open questions around the benefits or capabilities of any next-generation repository signing solutions (e.g. sigstore or the update framework (TUF))
+- Specific design or implementation strategies around binary signing (e.g. for Windows or macOS binaries)
+
+# Motivation
+[motivation]: #motivation
+
+The Rust Project currently lacks any official Public Key Infrastructure (PKI) required to perform proper package, mirror, code, and release signing across the project. There is high demand for capabilities that require such PKI infrastructure, such as mirroring and binary signing. This RFC provides the necessary cryptographic infrastructure to unblock such work, and an initial usage for cryptographic verification of Rust releases and `crates.io` downloads.
+
+We have chosen this specific approach to give us the most flexibility in choosing implementations for cryptographic security in future RFCs and implementations. Multiple problem spaces currently exist which can leverage an internal PKI - the most visible and obvious being crate and release signing. However, other avenues also exist - such as mirror validation or internal and external services communication. We see a future in which this Certificate Authority can be leveraged for meeting the needs of these problems, and consider it the first stepping stone towards a solution.
```

Comment:
As I mentioned multiple times [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/417663-tbd-signing/topic/PKI.20RFC.20-.20For.20Review), this RFC is doing two separate things:

* Defines the requirements for a Rust PKI around threats, quorum and key structure.
* Proposes implementing an X.509 CA to serve as the root of trust.

Implementing an X.509 CA *before* we decide on how we're going to sign stuff is premature, because if the signing scheme doesn't end up needing an X.509 CA we're going to have a ton of complication for little to no benefit.

This RFC should be split, moving the proposal for the X.509 CA to a separate RFC.

Reply:
To elaborate more, the concern of mine here is that this RFC mixes together how the PKI should work (having root keys, using a quorum to delegate, etc.) and that the PKI should be an X.509 CA, even though there is no accepted proposal that actually relies on X.509.

After a chat with @joshtriplett [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/417663-tbd-signing/topic/PKI.20RFC.20-.20For.20Review/near/423703010) I'm ok with not splitting the RFC, but that implies this RFC should not be accepted until an RFC using the X.509 PKI is accepted too, and this RFC might be rejected if we end up going in another direction. The split would just allow the implementation-agnostic parts to be accepted earlier.

Reply:
I would probably go further: I would treat it as a blocking concern to accept any particular proposal on even the properties (e.g., quorum structure, root keys, ...) without having a concrete use case in mind. I think we *could* accept an RFC that discusses the threat model, for example:

* We need a structure that is hard to compromise from one country

But going much beyond that is really hard without a concrete application. Even things like "what does revocation look like?" tie a lot into the specifics of what you are signing, when, and with what specific key. (And how clients are getting that key).

Reply:
@Mark-Simulacrum We can add more details about the threat model to the RFC.

Reply:
Addressed in the changes we just pushed, which include threat model details.